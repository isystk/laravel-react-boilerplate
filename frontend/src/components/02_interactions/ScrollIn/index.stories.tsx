import { Meta, Story } from '@storybook/react'
import React from 'react'
import ScrollIn, { ScrollInProps } from './index'

export default {
  title: '02_interactions/ScrollIn',
  component: ScrollIn,
} as Meta

const Template: Story = () => {
  const props: ScrollInProps = {
    delay: '1s',
  }
  return (
    <ScrollIn {...props}>
      <div>Hello!</div>
    </ScrollIn>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
