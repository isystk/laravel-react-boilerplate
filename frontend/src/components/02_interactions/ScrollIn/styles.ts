import { css, keyframes } from '@emotion/css'

export const fadeIn = css`
  opacity: 0;
  transform: translate(0, 50px);
`

const fadein = keyframes`
0% {
opacity: 0;
transform: translate(0, 50px);
}
100% {
opacity: 1;
transform: translate(0, 0);
}
`

export const scrollin = css`
  animation: ${fadein} 0.5s both;
`

export const delay = (props) => css`
  animation-delay: ${props.delay || '0s'};
`
