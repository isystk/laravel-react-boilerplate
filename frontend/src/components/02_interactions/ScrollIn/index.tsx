import React, { FC, useEffect, useRef, useState } from 'react'
import { ContainerProps, WithChildren } from 'types'
import { connect } from '@/components/hoc'
import * as styles from './styles'

/** ScrollInProps Props */
export type ScrollInProps = WithChildren & {
  className?: string
  delay?: string
}
/** Presenter Props */
export type PresenterProps = ScrollInProps

/** Presenter Component */
const ScrollInPresenter: FC<PresenterProps> = ({
  children,
  className,
  delay,
  myRef,
  isVisible,
  ...props
}) => (
  <div
    ref={myRef}
    className={`${className} ${
      isVisible ? styles.scrollin : styles.fadeIn
    } ${styles.delay({ delay })}`}
  >
    {children}
  </div>
)

/** Container Component */
const ScrollInContainer: React.FC<
  ContainerProps<ScrollInProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  const myRef = useRef(null)
  const [scroll, setScroll] = useState(0)
  const [windowHeight, setWindowHeight] = useState(0)
  const [isVisible, setIsVisible] = useState(false)

  const onScroll = (): void => {
    setScroll(window.scrollY || document.documentElement.scrollTop)
    setWindowHeight(window.innerHeight)
  }

  useEffect(() => {
    window.addEventListener('scroll', onScroll)
    onScroll()
  }, [])

  if (
    !isVisible &&
    myRef.current &&
    scroll > myRef.current.offsetTop - windowHeight + 100
  ) {
    setIsVisible(true)
    // 表示が完了したらスクロールイベントを削除
    window.removeEventListener('scroll', onScroll)
  }

  return presenter({
    children,
    myRef,
    isVisible,
    ...props,
  })
}

export default connect<ScrollInProps, PresenterProps>(
  'ScrollIn',
  ScrollInPresenter,
  ScrollInContainer
)
