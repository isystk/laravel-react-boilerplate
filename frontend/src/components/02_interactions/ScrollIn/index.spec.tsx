import React from 'react'
import renderer from 'react-test-renderer'
import ScrollIn, { ScrollInProps } from './index'

describe('ScrollIn', () => {
  it('Match Snapshot', () => {
    const props: ScrollInProps = {
      delay: '1s',
    }
    const component = renderer.create(
      <ScrollIn {...props}>
        <div>Hello!</div>
      </ScrollIn>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
