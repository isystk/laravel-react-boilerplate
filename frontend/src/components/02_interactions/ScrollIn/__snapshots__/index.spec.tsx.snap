// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollIn Match Snapshot 1`] = `
<div
  className="undefined css-pjt9eq css-xrumgy"
>
  <div>
    Hello!
  </div>
</div>
`;
